### PR TITLE
📬 Feature: post processors

### LINKED ISSUE #154: Expand on beautify output options
There's an undocumented feature that within a page.json you can beautify HTML output using [js-beauitfy](https://www.npmjs.com/package/js-beautify).

```JSON
{
  "settings": {
    "beautify": true
  }
}
```

This is not something you can set globally in a config.json, nor can it currently handle other outputs (XML, JS, JSON etc). I propose expanding the functionality and deprecating `dust.whitespace` so you can choose from minifying the output, beautifying it, or leaving it untouched in a single config option.

### PR DESCRIPTION
**Pull request by abovebored**
**Removes the previously undocumented `page.settings.beautify` option & closes #154 (related to #162).**

Adds the ability to manipulate the raw output of the template engine, before the page is cached and served. Similar to the [Wordpress](https://codex.wordpress.org/How_WordPress_Processes_Post_Content) feature, it is useful for linting (thinking RSS feeds in particular), minifying and formatting. They are loaded in order of definition, global first, and the output is passed from one to the next so you can chain functions.

Example usage: minifying HTML.

### config.json

```json
"globalPostProcessors": [
    "minify-html"
]
```

> *The default location in `paths.processors` is `workspace/processors`*

### `/workspace/processors/minify-html.js` (using [html-minifier](https://www.npmjs.com/package/html-minifier))

```js
const minify = require('html-minifier').minify

module.exports = (data, output) => {
  return minify(output, {
  	collapseWhitespace: true,
  	minifyCSS: true,
  	minifyJS: true,
  	removeRedundantAttributes: true,
  	useShortDoctype: true,
  	removeAttributeQuotes: true
  })
}
```

Note:

* the page json data object is also passed
* in addition it can be defined or disabled at page level in the `page.settings`

### Add a processor, in addition to the global

```json
"settings": {
  "postProcessors": ["remove-swear-words"]
}
 ```

### Disable for given page

```json
"settings": {
  "postProcessors": false
}
 ```

### Install package

```
npm install js-beautify --save
```

### New file `workspace/processors/beautify-html.js`

```js
const beautifyHtml = require('js-beautify').html

module.exports = (data, output) => {
  return beautifyHtml(output)
}
```

### Old `page.json`

```json
{
  "page": {
    "name": "Sitemap page",
    "description": "Sitemap",
    "language": "en"
  },
  "settings": {
    "cache": true,
    "beautify": true
  },
  ...
}
```

### New `page.json`

```json
{
  "page": {
    "name": "Sitemap page",
    "description": "Sitemap",
    "language": "en"
  },
  "settings": {
    "cache": true,
    "postProcessors": [
      "beautify-html"
     ]
  },
  ...
}
```

### CODE REVIEW

**Comment by jimlambie** [comment]
This is a breaking change because it removes `beautify`ing from core?

**Comment by abovebored** [comment]
Yes exactly, that setting will no longer have any effect...not sure how to handle it since it was undocumented anyway? You could only turn it on page-by-page in the `settings` block.

AFAIK we're not using it in any client work either.

**Comment by jimlambie** [comment]
Perhaps we should make some enquiries then - if it's not used we can simply make it a new feature and release as a minor version

**Comment by abovebored** [comment]
Added upgrade notes from `settings.beautify`

File: dadi/lib/view/index.js [hunk 1]
```diff
@@ -28,40 +27,48 @@ View.prototype.setData = function (data) {
...
+  if (this.json) return done(null, this.data)
 
-    this.template
-      .render(templateData, {
-        keepWhitespace: this.page.keepWhitespace
-      })
-      .then(output => {
-        let err = null
+  // Send the templated page
+  let templateData = Object.assign(this.data, {
```
Comment (eduardoboucas) [inline dadi/lib/view/index.js]: Could this be a `const`?

File: dadi/lib/view/index.js [hunk 2]
```diff
@@ -28,40 +27,48 @@ View.prototype.setData = function (data) {
...
+  this.template
+    .render(templateData, {
+      keepWhitespace: this.page.keepWhitespace
+    })
+    .then(output => {
+      let err = null
+
+      // Post-process the output
+      if (config.get('globalPostProcessors') || this.page.postProcessors) {
+        let postProcessors = config
```
Comment (eduardoboucas) [inline dadi/lib/view/index.js]: Could this be a `const`?

File: dadi/lib/view/index.js [hunk 3]
```diff
@@ -28,40 +27,48 @@ View.prototype.setData = function (data) {
...
+    host: this.page.hostKey
+  })
 
-        if (this.page.beautify) {
+  this.template
+    .render(templateData, {
+      keepWhitespace: this.page.keepWhitespace
+    })
+    .then(output => {
+      let err = null
```
Comment (eduardoboucas) [inline dadi/lib/view/index.js]: This variable doesn't seem to ever be changed. You're returning an error on L60 but it comes from L59. Could we remove this initialisation and just do `return done(null, output)` on L65?

File: dadi/lib/view/index.js (hunk 1 above)
Comment (eduardoboucas) [reply dadi/lib/view/index.js]: Also, to avoid mutating `this.data`, you could do `Object.assign({}, this.data, (...))`.

**Comment by abovebored** [comment]
Great suggestions, thanks @eduardoboucas!

**Comment by jimlambie** [comment]
This still has to be a major release, right? If we're not bundling in a beautify processor and honouring the existing page setting for beautify, then it's a breaking change. This would make it Web 5.0.

**Comment by abovebored** [comment]
Yep makes sense 👍

**Comment by eduardoboucas** [comment]
Agree 💯 %.

**Comment by jimlambie** [comment]
I'm a bit wary of another major release so soon after 4.0, is all. Feels like we're following the FB Messenger app approach - they're up to version 133 or something. Mind you, their release notes are awful. "Performance enhancements".

**Comment by eduardoboucas** [comment]
That's understandable. If this is not an urgent requirement, perhaps we can hit pause and bundle it with other breaking changes for the next major release? ⏸

**Comment by abovebored** [comment]
Could we prepare a 5.0 branch perhaps? Just conscience that this branch will get stale and drift from head.

**Comment by eduardoboucas** [comment]
Good idea. We could start a `release/5.0` branch – I think this is the convention we agreed on, right @jimlambie?

**Pull request merged**